[#2] Header 구현

Huon0423 opened the pull request:
## 개요
- 로그인 상태 전역 관리를 위해 zustand를 사용
- tailwind와 headlessui를 사용하여 회원 Header UI 컴포넌트 및 비회원 Header UI 컴포넌트 구현
  (tailwind의 dropdown이 안되는 문제로 headlessui의 dropdown 사용)
- react-router-dom으로 메뉴 라우팅 설정
- 회원 / 비회원 전환 확인을 위한 일부 코드가 같이 들어가있어 추후 수정 필요

## 변경 사항

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.

Huon0423 committed:
Feat: [FE] Header 구현

mj0107 commented on src/store/store.ts:
추후에 다른 store 생성시 다른 store와의 혼동 가능성이 있습니다. 파일명을 명확히 바꿔주세요

Huon0423 replied:
store.ts에 기능별로 export를 여러개 넣을 가능성이 있겠다 싶었는데 그냥 따로 해당 기능 관련 store 파일을 만드는게 낫겠군요. 바꾸겠습니당

mj0107 commented on src/store/store.ts:
```diff
@@ -0,0 +1,9 @@
+import { create } from "zustand";
+import { AuthState } from "@/types/user";
+
+export const useAuthStore = create<AuthState>((set) => ({
```
그냥 전역상태로 저장하면 isLogin과 user는 새로고침시 초기화가 됩니다.
web storage를 이용해서 저장해주시는게 어떨까요

Huon0423 replied:
헐,, 관련해서 알아보고 수정하겠습니다,,

mj0107 commented on src/types/user.ts:
isLogin과 user는 state가 맞지만,
login과 logout은 함수입니다. 따라서 action이 됩니다.

state와 action을 분리하여 타입지정을 하는건 어떨까요

Huon0423 replied:
ts user interface 관련해서 알아볼 때 action이 따로 있는걸 못 찾아봤네요. 찾아서 더 공부하고 분리해두겠습니당,,,

mj0107 replied:
> ts user interface 관련해서 알아볼 때 action이 따로 있는걸 못 찾아봤네요. 찾아서 더 공부하고 분리해두겠습니당,,,

ts에 관련된 피드백이 아니라 개념에 관련된 피드백이었습니다.
state는 상태를 나타냅니다. 흔히 react의 state를 생각하면 됩니다.
**action은 그 상태를 변화시키는 함수들을 일컫습니다.**

아래는 제가 로그인 전역상태를 구현할 때 사용했던 코드입니다.
```ts
import { create } from 'zustand'
import { createJSONStorage, persist } from 'zustand/middleware'

interface authState {
  isSignedIn: boolean
}
interface authAction {
  setSignedIn: () => void
  setSignedOut: () => void
}

export const useAuthStore = create<authState & authAction>()(
  persist(
    (set) => ({
      isSignedIn: false,
      setSignedIn: () => set({ isSignedIn: true }),
      setSignedOut: () => set({ isSignedIn: false }),
    }),
    {
      name: 'isSignedIn',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
```

아래는 공식문서에 추천하는 패턴입니다.
https://docs.pmnd.rs/zustand/guides/practice-with-no-store-actions

참고하시길 바랍니다.

Huon0423 replied:
감사합니다 :)

mj0107 commented on src/layouts/Header.tsx:
```diff
@@ -0,0 +1,65 @@
+      <nav className="flex items-center justify-between px-6 py-1">
+        <NavLink to={"/"}>
+          <img src="/brand.svg" alt="brand" className="h-7 w-auto"></img>
+        </NavLink>
```
semantic 구조에 따라 해당부분은 nav 밖으로 빼는게 나을것 같은데 어떻게 생각하시나요

Huon0423 replied:
NavLink를 사용하기에 Nav안에 넣어뒀는데, Nav는 Header의 navigation 메뉴이니 클릭시 홈으로 이동하더라도 로고는 Nav 밖으로 빼는게 맞는 거 같습니다. 수정하겠습니다.

mj0107 commented on src/layouts/Header.tsx:
```diff
@@ -0,0 +1,65 @@
+              <MenuItem>
+                <div className="px-1 py-1">
+                  {/* <NavLink to={"/login"}>로그인</NavLink> */}
+                  <div onClick={handleLogin}>로그인</div>
```
이부분은 왜 div로 하셨나요?

Huon0423 replied:
로그인시 원래 모달창이 떠야하는데 NavLink주석 처리부터 틀렸군요,, div는 일시적으로 로그인 / 로그아웃 변화를 체크하기 위해 임의로 넣은 거라 편의상 div로 넣었습니다.  48번째 줄의 ```<div> onClick={logout}>로그아웃</div>``` 도 div로 되어 있긴 한데 별 이유가 없었습니다. hover했을때 커서가 text 형태 그대로이니 button으로 바꾸는게 나을까요?

mj0107 replied:
가능한 설명을 위한 간단한 주석 외에 테스트나 임시를 위한 주석은 빼주시고 올려주세요

Huon0423 replied:
알겠습니다!

Huon0423 committed:
Refactor : [FE] Header authStore 수정

- store 간의 혼동 가능성 => store.ts에서 authStore.ts로 이름 수정
- 재사용되지 않는 user.ts를 제거하고 authStore.ts에 삽입
- auth의 store에서 state와 action 분리
- 수정한 authStore에 맞춰 header 수정

Huon0423 committed:
Refactor : [FE] header.tsx의 html 요소 이동 및 변경

- semantic구조에 따라 logo의 위치를 nav 밖으로 이동
- 로그인 및 로그아웃 버튼 div에서 button으로 변경
- 임시 주석 제거

Huon0423 committed:
Feat : [FE] Header persist 추가

- 새로고침시 초기화되지 않도록 useAuthStore에 persist 추가
- persist 추가에 따라 useAuthStore에서 빼놓은 함수를 다시 내부로 삽입
- Header 코드를 그에 맞게 일부 수정

mj0107 commented on src/App.tsx:
```diff
@@ -1,5 +1,6 @@
-import { Outlet } from 'react-router-dom';
+import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
+import { Outlet } from "react-router-dom";
+import Header from "./layouts/Header";
```
alias를 이용한 절대경로로 바꿔주세요

Huon0423 replied:
혹시 절대경로를 사용해야하는 이유가 가독성 외에 더 있나요?

mj0107 replied:
> 혹시 절대경로를 사용해야하는 이유가 가독성 외에 더 있나요?

일종의 컨벤션이라고 볼 수 있고, 통일성때문에 그렇습니다.
위는 바로 상위 디렉토리를 참조하는 것이라 ./만 들어가는데, 2~3개 밖에 있는 파일이나 컴포넌트를 참조하게 될 경우
../../../와 같이 가독성을 해치는 이유도 있습니다.

원래는 assets, components, pages 등등의 디렉토리를 @assets, @components, @pages와 같이 모두 alias 설정을 해주지만, 지금은 루트 디렉토리에 대한 @ alias 설정만 있으니 통일성을 위해 사용합니다. 추후에 alias는 구조가 잡힌후에 추가하도록 하겠습니다.

Huon0423 replied:
아하 알겠습니다! 감사합니다 :)

mj0107 commented on src/layouts/Header.tsx:
```diff
@@ -0,0 +1,66 @@
+                  </div>
+                </MenuItem>
+                <MenuItem>
+                  <div className="px-1 pb-1">
```
button을 div로 감싼이유가 뭔가요

Huon0423 replied:
NavLink일 때 만든 div를 지우는걸 깜빡했습니다.. 지우겠습니다!
